tools/releases/changelog/root.go

```diff
@@ -121,51 +79,64 @@ var graphqlCmd = &cobra.Command{
 		}
 
 		// Rollup changes together
-		byChangelog := map[string]*Changelog{}
+		byChangelog := map[string][]*CommitInfo{}
 		for i := range res {
 			if strings.HasPrefix(commitLimit, res[i].Oid) {
 				break
 			}
 			ci := NewCommitInfo(res[i])
-			if ci.ShouldIgnore() {
+			if ci == nil {
 				continue
 			}
-			cl := ci.Changelog
-			if ci.Changelog == "" {
-				cl = ci.PrTitle
-			}
-			c := byChangelog[cl]
-			if c == nil {
-				c = &Changelog{
-					Desc:         cl,
-					PullRequests: []int{},
+			byChangelog[ci.Changelog] = append(byChangelog[ci.Changelog], ci)
+		}
+		// Create a list to display
+		var out []Changelog
+		for changelog, commits := range byChangelog {
+			uniqueAuthors := map[string]interface{}{}
+			var authors []string
+			var prs []int
+			for _, c := range commits {
+				prs = append(prs, c.PrNumber)
+				if _, exists := uniqueAuthors[c.Author]; !exists {
+					authors = append(authors, fmt.Sprintf("@%s", c.Author))
+					uniqueAuthors[c.Author] = nil
 				}
-				byChangelog[cl] = c
 			}
```

Comment:
Question: I have used to represent sets as `map[string]struct{}` and not `map[string]interface{}` as `struct{}`s are not taking any space in memory. My assumption is that when compiler is generating machine code, it knows looking at the type of values (`struct{}`) how much memory it will need to prepare. My assumption is that when using `map[string]interface{}` it makes compiler harder to asses the necessary space (just my assumption, not confirmed empirically). Is there any reason I haven't though about for using `map[string]interface{}`?

Reply:
Ok, I compared the results and `map[string]struct{}` generates smaller intermediate representation
```
genssa

    # /Users/bartsmykla/Projects/github.com/bartsmykla/go-playground/main.go
    00000 (7) TEXT "".main(SB), ABIInternal
    00001 (7) FUNCDATA $0, gclocals·69c1753bd5f81501d95132d08af04464(SB)
    00002 (7) FUNCDATA $1, gclocals·713abd6cdf5e052e4dcd3eb297c82601(SB)
    00003 (7) FUNCDATA $2, "".main.stkobj(SB)
v4
    00004 (+8) PCDATA $1, $0
v4
    00005 (+8) CALL runtime.makemap_small(SB)
v18
    00006 (8) MOVQ AX, "".foo-24(SP)
v32
    00007 (+9) MOVQ AX, BX
v24
    00008 (9) LEAQ go.string."1"(SB), CX
v33
    00009 (9) MOVL $1, DI
v23
    00010 (9) LEAQ type.map[string]struct {}(SB), AX
v10
    00011 (+9) PCDATA $1, $1
v10
    00012 (+9) CALL runtime.mapassign_faststr(SB)
v13
    00013 (9) TESTB AX, (AX)
v46
    00014 (+12) MOVUPS X15, ""..autotmp_13-16(SP)
v26
    00015 (12) LEAQ type.map[string]struct {}(SB), AX
v35
    00016 (12) MOVQ AX, ""..autotmp_13-16(SP)
v44
    00017 (12) MOVQ "".foo-24(SP), AX
v31
    00018 (12) MOVQ AX, ""..autotmp_13-8(SP)
v37
    00019 (?) NOP
    # $GOROOT/src/fmt/print.go
v40
    00020 (+274) MOVQ os.Stdout(SB), BX
v45
    00021 (274) LEAQ go.itab.*os.File,io.Writer(SB), AX
v39
    00022 (274) LEAQ ""..autotmp_13-16(SP), CX
v21
    00023 (274) MOVL $1, DI
v41
    00024 (274) MOVQ DI, SI
v42
    00025 (274) PCDATA $1, $0
v42
    00026 (274) CALL fmt.Fprintln(SB)
    # /Users/bartsmykla/Projects/github.com/bartsmykla/go-playground/main.go
b4
    00027 (13) RET
    00028 (?) END
```
for code:
```go
package main

import (
	"fmt"
)

func main() {
	foo := map[string]struct{}{
		"1": {},
	}

	fmt.Println(foo)
}
```

than `map[string]interface{}`
```
    # /Users/bartsmykla/Projects/github.com/bartsmykla/go-playground/main.go
    00000 (7) TEXT "".main(SB), ABIInternal
    00001 (7) FUNCDATA $0, gclocals·69c1753bd5f81501d95132d08af04464(SB)
    00002 (7) FUNCDATA $1, gclocals·713abd6cdf5e052e4dcd3eb297c82601(SB)
    00003 (7) FUNCDATA $2, "".main.stkobj(SB)
v4
    00004 (+8) PCDATA $1, $0
v4
    00005 (+8) CALL runtime.makemap_small(SB)
v21
    00006 (8) MOVQ AX, "".foo-24(SP)
v36
    00007 (+9) MOVQ AX, BX
v29
    00008 (9) LEAQ go.string."1"(SB), CX
v27
    00009 (9) MOVL $1, DI
v26
    00010 (9) LEAQ type.map[string]interface {}(SB), AX
v10
    00011 (+9) PCDATA $1, $1
v10
    00012 (+9) CALL runtime.mapassign_faststr(SB)
v16
    00013 (9) MOVQ $0, (AX)
v25
    00014 (9) PCDATA $0, $-2
v25
    00015 (9) CMPL runtime.writeBarrier(SB), $0
b4
    00016 (9) JNE 19
v22
    00017 (9) MOVQ $0, 8(AX)
b3
    00018 (9) JMP 22
v17
    00019 (9) LEAQ 8(AX), DI
v15
    00020 (9) XORL DX, DX
v30
    00021 (9) CALL runtime.gcWriteBarrierDX(SB)
v49
    00022 (12) PCDATA $0, $-1
v49
    00023 (+12) MOVUPS X15, ""..autotmp_13-16(SP)
v18
    00024 (12) LEAQ type.map[string]interface {}(SB), DX
v38
    00025 (12) MOVQ DX, ""..autotmp_13-16(SP)
v31
    00026 (12) MOVQ "".foo-24(SP), DX
v34
    00027 (12) MOVQ DX, ""..autotmp_13-8(SP)
v40
    00028 (?) NOP
    # $GOROOT/src/fmt/print.go
v43
    00029 (+274) MOVQ os.Stdout(SB), BX
v37
    00030 (274) LEAQ go.itab.*os.File,io.Writer(SB), AX
v44
    00031 (274) LEAQ ""..autotmp_13-16(SP), CX
v24
    00032 (274) MOVL $1, DI
v35
    00033 (274) MOVQ DI, SI
v45
    00034 (274) PCDATA $1, $0
v45
    00035 (274) CALL fmt.Fprintln(SB)
    # /Users/bartsmykla/Projects/github.com/bartsmykla/go-playground/main.go
b2
    00036 (13) RET
    00037 (?) END
```

for code:
```go
package main

import (
	"fmt"
)

func main() {
	foo := map[string]interface{}{
		"1": nil,
	}

	fmt.Println(foo)
}
```

Reply:
Yes my bad I meant to use struct :)

Reply:
OK @bartsmykla can you benchmark that? :grinning: For me the biggest reason to use `map[string]struct{}` is just the clear semantics, that the type `map[string]struct{}` is the only type that's isomorphic to a theoretical `set[string]`

Reply:
```sh
goos: darwin
goarch: amd64
pkg: go-playground
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkMapOfStructsTest
BenchmarkMapOfStructsTest-12       	      16	  64450236 ns/op
BenchmarkMapOfInterfacesTest
BenchmarkMapOfInterfacesTest-12    	      12	  89367545 ns/op
PASS
```